### PR TITLE
[@types/meteor] Add types for meteor/reactive-dict.

### PR DESCRIPTION
**Pull request by szchenghuang**
Type definitions for meteor/reactive-dict.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [ ] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/meteor/globals/reactive-dict.d.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+declare var ReactiveDict: ReactiveDictStatic;
+declare interface ReactiveDictStatic {
+    new (name?: string, initialValue?: EJSONable): ReactiveDict;
+}
+declare interface ReactiveDict {
+    setDefault(key: string, value: EJSONableProperty): void;
+    setDefault(object: EJSONableProperty): void;
+    set(key: string, value: EJSONableProperty): void;
```
Comment (radekmie) [inline types/meteor/globals/reactive-dict.d.ts:8]: Here and in `setDefault(key: string, value: EJSONableProperty): void` the `value` should be optional. [Source](https://github.com/meteor/meteor/blob/71f67d9dbac34aba34ceef733b2b51c2bd44b665/packages/reactive-dict/reactive-dict.js#L82).

File: types/meteor/globals/reactive-dict.d.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    set(object: EJSONableProperty): void;
+    get(key: string): EJSONableProperty;
+    equals(key: string, value: string | number | boolean | undefined): boolean;
+    all(): EJSONable;
+    clear(): void;
+    destroy(): boolean;
```
Comment (radekmie) [inline types/meteor/globals/reactive-dict.d.ts:14]: It should have a `void` return type. [Source](https://github.com/meteor/meteor/blob/71f67d9dbac34aba34ceef733b2b51c2bd44b665/packages/reactive-dict/reactive-dict.js#L285-L298). There is also `delete(key: string): boolean` ([source](https://github.com/meteor/meteor/blob/71f67d9dbac34aba34ceef733b2b51c2bd44b665/packages/reactive-dict/reactive-dict.js#L260-L283)), but it's not in [Meteor Docs](https://docs.meteor.com/api/reactive-dict.html), therefore I'd leave it for now.

**Review by radekmie — CHANGES_REQUESTED** [review]

File: types/meteor/globals/reactive-dict.d.ts (hunk 1 above)
Comment (szchenghuang) [reply types/meteor/globals/reactive-dict.d.ts]: @radekmie Both methods are overloaded. Do you prefer not?

```
    setDefault(key: string, value: EJSONableProperty): EJSONable;
    setDefault(object: EJSONableProperty): EJSONable;
               ^^^^^^^^^^^^^^^^^^^^^^^^^

    set(key: string, value: EJSONableProperty): EJSONable;
    set(object: EJSONableProperty): EJSONable;
               ^^^^^^^^^^^^^^^^^^
```

File: types/meteor/globals/reactive-dict.d.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+declare var ReactiveDict: ReactiveDictStatic;
+declare interface ReactiveDictStatic {
+    new (name?: string, initialValue?: EJSONable): ReactiveDict;
+}
+declare interface ReactiveDict {
+    setDefault(key: string, value: EJSONableProperty): void;
+    setDefault(object: EJSONableProperty): void;
+    set(key: string, value: EJSONableProperty): void;
+    set(object: EJSONableProperty): void;
```
[the remark this answers is not in the document]
Comment (radekmie) [reply types/meteor/globals/reactive-dict.d.ts]: Sorry for an ambiguous comment @szchenghuang - I meant the argument type, not the return type.

File: types/meteor/globals/reactive-dict.d.ts (hunk 1 above)
Comment (radekmie) [reply types/meteor/globals/reactive-dict.d.ts]: They should be overloaded, but in the first case the `value` is optional (or at least it seems like it in the [source](https://github.com/meteor/meteor/blob/71f67d9dbac34aba34ceef733b2b51c2bd44b665/packages/reactive-dict/reactive-dict.js#L82)):
```ts
    set(key: string, value?: EJSONableProperty): void;
    set(object: EJSONable): void;
```

Those won't collide as `EJSONable` is an object.

File: types/meteor/globals/reactive-dict.d.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (szchenghuang) [reply types/meteor/globals/reactive-dict.d.ts]: Cool. Changed the argument type and reverted the return one.

File: types/meteor/globals/reactive-dict.d.ts (hunk 1 above)
Comment (szchenghuang) [reply types/meteor/globals/reactive-dict.d.ts]: Let me know what else to polish @radekmie? Thanks!

**Review by radekmie — APPROVED** [review]

**Pull request merged**